docs/langref/relay_adt.rst

```diff
@@ -0,0 +1,517 @@
+
+.. code_block:: python
+
+   def @unfoldr<a, b>(%f : fn(b) -> Optional[(a, b)], %z : b) -> List[a] {
+     match(%f(%z)) {
+       case Some(%pair) { Cons(%pair.0, @unfoldr(%f, %pair.1)) }
+       case None() { Nil() }
+     }
+   }
+
+   # we need some way of generating an input to the cell function given only a state
+   def @gen_func<state_type, in_type, out_type>(%state : state_type) : Optional[(out_type, state_type)] {
+     let %in : Optional[in_type] = @generate_input(%state);
+     match(%in) {
+       case Some(%n) {
+         let %cell_out = @cell(%n, %state);
+         Some((%cell_out.1, %cell_out.0)) # pair of output and state
+       }
+       case None() { None() }
+     }
+   }
+
+   def @generator<state_type, in_type, out_type>(%cell, %init : state_type) -> List[out_type] {
+     @unfoldr(fn(%state) { @gen_func(%cell, %state) }, %init)
+   }
+
+An accumulating map (a fold that simultaneously updates an accumulator value and a list
+of outputs) can be used to write a general RNN (with an output for every input):
+
+.. code_block:: python
+
+   def @map_accumr<a, b, c>(%f : fn(a, b) -> (a, c), %acc : a, %l : List[b]) -> (a, List[c]) {
+     match(%l) {
+       case Nil() { (%acc, Nil()) }
+       case Cons(%b, %t) {
+         let %update = %f(%acc, %b);
+         let %rest = @map_accumr(%f, %update.0, %t));
+         (%rest.0, Cons(%update.1, %rest.1))
+       }
+     }
+   }
+
+   # can also be implemented as a right fold:
+   def @map_accumr_fold(%f, %acc, %l) {
+     @foldr(fn(%b, %p) {
+       let %f_out = %f(%p.0, %b);
+       (%f_out.0, Cons(%f_out.1, %p.1))
+     },
+     (%acc, Nil()), %l)
+   }
+
+   def @general_rnn<state_type, in_type, out_type>(%cell, %init : state_type, %input : List[in_type])
+     -> (state_type, List[out_type]) {
+     @map_accumr(%cell, %init, %input)
+   }
+
+Olah also gives an example of a bidirectional neural network, in which two sets of
+cells (which may have different weights) process the input in both directions and produce a
+single set of outputs. The following is a Relay implementation of that example:
+
+.. code_block:: python
+
+   # creates a list of tuples from two lists
+   def @zip<a, b>(%l : List[a], %m : List[b]) -> List[(a, b)] {
+     match(%l) {
+       case Nil() { Nil() }
+       case Cons(%a, %t1) {
+         match(%m) {
+           case Nil() { Nil() }
+           case Cons(%b, %t2) { Cons((%a, %b), @zip(%t1, %t2)) }
+         }
+       }
+     }
+   }
+
+   # analogous to map_accumr
+   def @map_accmul(%f, %acc, %l) {
+     @foldl(fn(%p, %b){
+       let %f_out = %f(%p.0, %b);
+       (%f_out.0, Cons(%f_out.1, %p.1))
+     }, (%acc, Nil()), %l)
+   }
+
+   def @bidirectional_rnn<state1_type, state2_type, in_type, out1_type, out2_type>
+     (%cell1, %cell2, %state1 : state1_type, %state2 : state2_type, %input : List[in_type])
+     -> List[(out1_type, out2_type)] {
+     @zip(@map_accumr(%cell1, %state1, %input).1, @map_accuml(%cell2, %state2, %input).1)
```

Comment:
[BRNN](https://www.coursera.org/lecture/nlp-sequence-models/bidirectional-rnn-fyXnn) also applies an activation function on the output of `zip`. Maybe change it to `map linear zip(..)` .

Reply:
Don't do that. Output projection is not always required.

Reply:
ok. let's skip this comment :-)

Reply:
Noted